src/pybind/mgr/coleoid/module.py

```diff
@@ -0,0 +1,432 @@
+"""
+A RESTful API for Ceph
+"""
+
+import json
+import inspect
+import StringIO
+import threading
+import traceback
+import ConfigParser
+
+import api
+import common
+
+from uuid import uuid4
+from flask import Flask
+from OpenSSL import SSL
+from flask_restful import Api
+
+from mgr_module import MgrModule, CommandResult
+
+# Global instance to share
+instance = None
+
+
+
+class CommandsRequest(object):
+    """
+    This class handles parallel as well as sequential execution of
+    commands. The class accept a list of iterables that should be
+    executed sequentially. Each iterable can contain several commands
+    that can be executed in parallel.
+
+    Example:
+    [[c1,c2],[c3,c4]]
+     - run c1 and c2 in parallel
+     - wait for them to finish
+     - run c3 and c4 in parallel
+     - wait for them to finish
+    """
+
+
+    @staticmethod
+    def run(commands, uuid=str(uuid4())):
+        """
+        A static method that will execute the given list of commands in
+        parallel and will return the list of command results.
+        """
+
+        # Gather the results (in parallel)
+        results = []
+        for index in range(len(commands)):
+            tag = '%s:%d' % (str(uuid), index)
+
+            # Store the result
+            result = CommandResult(tag)
+            result.command = common.HUMANIFY_COMMAND(commands[index])
+            results.append(result)
+
+            # Run the command
+            instance.send_command(result, json.dumps(commands[index]), tag)
+
+        return results
+
+
+    def __init__(self, commands_arrays):
+        self.uuid = str(id(self))
+
+        # Filter out empty sub-requests
+        commands_arrays = filter(
+            lambda x: len(x) != 0,
+            commands_arrays,
+        )
+
+        self.running = []
+        self.waiting = commands_arrays[1:]
+        self.finished = []
+        self.failed = []
+
+        self.lock = threading.RLock()
+        if not len(commands_arrays):
+            # Nothing to run
+            return
+
+        # Process first iteration of commands_arrays in parallel
+        try:
+            results = self.run(commands_arrays[0], self.uuid)
+        except:
+            instance.log.error(str(traceback.format_exc()))
```

Comment:
Looks like I forgot to remove this debugging try: except: statement which should no longer be necessary (I have previously had some issues with logging from different places so I had to catch this early but it is no longer necessary).